### PR TITLE
"headless" support for VMs (non-desktop linux instances) Issue #15

### PR DESCRIPTION
**Pull request by tonymet**
**DEPS** : check out https://github.com/tonymet/oauth2-noserver/commits/feature/headless first

Addresses Issue #15 -- gphotos-uploader-cli will now run on a standard linux vm without gnome installed (tested on ubuntu & chromium)

* tests for dbus, and uses main leveldb (key = `credential_USERNAME` ) for credentials if dbus is not running .  Otherwise uses `zolando/go-keyring` as before.
* allows `redirectURL` to be set in config as `APIAppCredentials.RedirectURL` (needs `tonymet/feature/headless` branch of `oauth2-noserver`)

### Example config
```
{
  APIAppCredentials: {
    ClientID:     "XXXX"
    ClientSecret: "YYYYYY",
    RedirectURL:  "http://myserver.tonym.us/oauth/callback"
....
}
```

### CODE REVIEW

**Comment by rfgamaral** [comment]
If I'm not mistaken, this will not work for me for the same reason that I don't think this is the best approach for general use. I mean, this relies on a publicly exposed machine for the redirect URI, most users won't have that on their VMs, NAS, whatever...

Also, for easier reviewing and integration, I think it would help @nmrshll if you could do multiple pull requests with individual features instead of merging a bunch into a single PR. The way I see it, this PR could at least be separated into two different things (token storage for one, headless authentication as another).

My 2 cents.

**Comment by tonymet** [comment]
@rfgamaral good feedback.  I'd still appreciate your help testing if you have a few min. 

The work should be a step in the right direction -- there's no clear path to spinning up an external server that I'm aware of.  This will at least let you configure your own--which helped me a lot.

Let's chat with @nmrshll and see if he cares to split.  I'm ok either way, but the value as a story is better bound up this way IMO.

**Comment by rfgamaral** [comment]
> I'd still appreciate your help testing if you have a few min.

I don't have a local server, publicly exposed, running on my NAS. I can't help you test your full approach.

> The work should be a step in the right direction

Please don't take this in the wrong way but I only partially agree. Yes, storing the token in a more cross-platform way (at least not tied to gnome) is a step in the right direction. But I don't agree with the other bit, setting up a local server (exposed and accessible externally). I don't think that's the way forward.

> Let's chat with @nmrshll and see if he cares to split. I'm ok either way, but the value as a story is better bound up this way IMO.

I think both features bring value individually and I'm personally more interested in getting that token work merged into `master` than the other bit (for the reasons mentioned above). One thing does not invalidate the other, the token work as a lot of value as a standalone feature.

**Comment by tonymet** [comment]
It's all good. 

About the nas, I've done this before by giving your nas a hostname (even in your hosts file or your domain) . Only your browser needs to reach it (not google).  the RedirectURL support will help facilitate this (e.g. use nas.rfgamaral.net ). 

I'm pretty sure the merge will be smooth since the RedirectURL changes are backward compatible and don't change the config (the default will use the old url).  

Moreover, the token storage is backward compatible as well.  

Long story short the new code will help you, and we can work on helping you with the external server if it's still needed.

**Comment by tonymet** [comment]
@rfgamaral btw i pinged you on skype to help discuss the solution skype=tonymetzynga

**Comment by rfgamaral** [comment]
You misunderstood me. I know how to setup a domain and an external server on my NAS and how to point that domain to my NAS. It's just something that I don't want to do, especially not for something like this.

This is a tool for uploading and just that. Uploading from my server to my Google Photos account. I don't want (nor need) my NAS to be publicly accessible just so that I can authenticate this tool. You have to understand this is not supposed - at least from my point of view - to just you and me or technically savvy people. We should cater to all types of users.

With that in mind, it's not about making things backwards compatible just for the sake of it. If it makes sense (for some reason), sure. If not, and if we can have one good solution that works for most (if not all) use cases we should go with that. Otherwise, we're just adding up to the code base, making it more complex and harder to maintain because we'll have multiple solutions to to achieve the same end result. That's usually not a very good idea.

As for "helping me", that's why (but not only of course) I suggested to make two separate PRs 😁

For my Docker image to be 100% usuable I need a few things:

1. Access token needs to be stored with your code and not in gnome keyring. If this were a separate PR it would be better for me so I wouldn't need to maintain my own fork of this project. It would also be better from a code review point of view and we could focus on discussing that implementation and perhaps discuss the possibility of replacing the token storage completely with your code instead of still relying on gnome keyring first. But as you see, with two changes in a single PR we have cross discussions of two features that bring value individually.
2. It needs an automated mechanism for renewal of the access token (by using the refresh token that it's not currently stored anyway as far as I know). I'm not sure if @nmrshll is working on this or has plans to but this is something required not only for my Docker image but for anyone that wants to run this on a headless server (otherwise periodical and manual authentication will be required).
3. A more seamless authentication process is required (the other bit you're working on this same PR). The way I see it, we really need to research and think this more thoroughly to come up with a more elegant and reliable solution that's as user friendly as possible. I'm personally "happy" with the current manual solution I've detailed on the README of my Docker image.

**Comment by rfgamaral** [comment]
Oh, and I'm not sure who you pinged but I don't use Skype 😆

**Comment by ManuCart** [comment]
Hi guys,
I don't really understood what you are doing
I have a Raspberry Pi used as headless server.
I allready use this scripts : 
````
https://github.com/odeke-em/drive
https://github.com/thebigmunch/google-music-scripts
https://github.com/tokland/youtube-upload
https://github.com/andreafabrizi/Dropbox-Uploader
```` 
 And the only thing to do is to copy OAuth Address to my browser and copy back token or credential to my headless server.
It's should be same simpler for google photo

Please help me to do this script work without docker

**Comment by tonymet** [comment]
@rfgamaral I think we should chat on skype it'll be more constructive. I think we can come up with a solution. 
About the nas, it doesn't have to be public, just have a name. e.g. "nas.rfgamal.com" in your hosts file on your laptop where you auth. 
The refresh token is saved in both cases.  If you are not seeing it, you'll have to de-auth the app (go to google->my account-> 3rd party apps -> delete).  Refresh tokens are only offered on the first auth and not subsequent auths.  The uploader uses golang's standard oauth lib which refreshes properly. 

@ManuCart the crux of the problem is that photoslibrary is different and doesn't support the copy-paste flow like those other tools.  It's a newer api and google is restricting the allowed flows in oauth2.  If someone wants to test this to double check my work please do and I'm happy to implement the copy-paste flow.

**Comment by tonymet** [comment]
@rfgamaral saw the note i'll try keybase

**Comment by ManuCart** [comment]
Ok, I will try.
What the way to install your work

**Comment by tonymet** [comment]
@ManuCart 
```
go get github.com/nmrshll/gphotos-uploader-cli
cd $GOPATH/src/github.com/nmrshll/oauth2-noserver
git remote add tonymet	git@github.com:tonymet/oauth2-noserver
git checkout feature/headless
cd $GOPATH/src/github.com/nmrshll/gphotos-uploader-cli
git remote add tonymet	git@github.com:tonymet/gphotos-uploader-cli.git
git checkout feature/headless-complete
make build
```

Update your config to add RedirectURL (use your hostname--this url must be added to your oauth credentials in your GCP account
```
{
  APIAppCredentials: {
    ClientID:     "XXXX"
    ClientSecret: "YYYYYY",
    RedirectURL:  "http://myserver.tonym.us/oauth/callback"
....
}
```

**Comment by tonymet** [comment]
@nmrshll this feels pretty stable and backward compatible.  part of the discussion focused on other ways to make headless easier, but this patch should definitely help people running non-desktop linux (e.g. ubuntu on a vm, or a nas installation). 

Could I ask for your feedback and help to get it merged in I think it's good work :D  . 

Thanks for all the support.

**Comment by rfgamaral** [comment]
@tonymet Not trying to diminish your work, it's awesome that some one with Go skills is helping fixing important issue like the headless one but I'm still a bit skeptical about this implementation. Please answer me this:

How can a user, without owning a domain like your `tonym.us`, can use this implementation to successfully authenticate this on headless server? Honest question, I'm not seeing how it works.

**Comment by tonymet** [comment]
Happy to answer questions always good to have feedback.  

First off the goal here is just to improve compatibility to make sure gphotos-uploader can run anywhere -- when I ran on docker or a vm it would crash due to 1. lack of dbus and (2) unable to launch a new window.  So now any linux can run. 

About running without DNS it is tricky but doable.  Of course the easiest would be to buy a $2 domain and then put your VM's public IP on gphotos.mydomain.com .   In lieu of that, since 127.0.0.1:8080/oauth/callback is allowed, one could `ssh -L8080:localhost:14656  ec2-user@vm-public-ip`  in order to forward the oauth redirect to gphotos-uploader-cli on a remote vm.  

I'm happy to write this up and add to the documentation as a possible workaround.

**Comment by rfgamaral** [comment]
@tonymet Tried to [cherry-pick your credentials related changes](https://github.com/rfgamaral/gphotos-uploader-cli/commit/931fbd8abf88f71f5a7a9d470728c97713a93b21) from this PR into my own fork (that I use on my Docker image) and got the following error compiling `gphotos-uploader-cli`:

```
# github.com/nmrshll/gphotos-uploader-cli/upload
../../upload/folderUploadJob.go:49:5: undefined: tokenstore
../../upload/folderUploadJob.go:51:3: undefined: tokenstore
../../upload/folderUploadJob.go:54:3: undefined: tokenstore
../../upload/folderUploadJob.go:68:16: undefined: tokenstore
../../upload/folderUploadJob.go:91:8: undefined: tokenstore
```

At first I thought I forgot to cherry-pick something important but looking at the changes on this PR itself, it looks to me that you're [missing an import statement](https://github.com/rfgamaral/gphotos-uploader-cli/commit/8e12039ff8f2f33f78ca11df63dfbe7728e0b363) on `folderUploadJob.go` to use `tokenstore`? After adding that change to my fork I was able to compile successfully, is this the right way to fix this problem?

**Comment by tonymet** [comment]
@rfgamaral thanks for checking.  It seems like it may be an issue with the cherry pick. I did a clean checkout and reviewed the diff (see the files tab).  There are a couple commits here so they may have been applied out of order.

**Comment by tonymet** [comment]
@rfgamaral to help you out I went ahead and squashed this branch down to one commit.  You'll need to reset your local branch here's how.  then you can cherry-pick more easily

```
git checkout feature/headless-complete
git reset --hard tonymet/feature/headless-complete
# double check
git log -1
4d36ae0 - (HEAD -> feature/headless-complete, tonymet/feature/headless-complete) configure RedirectURL in config.hjson. add TokenStoreLevelDB -- support running w/o dbus (2 minutes ago) <Anthony Metzidis>
```

**Comment by rfgamaral** [comment]
Well, I'm not sure what's going on, I have most of your code on my fork, the only two files I didn't picked were:

- `.gitignore`
- `config.go`

I still don't understand why I wasn't able to compile your branch without adding the import statement I mentioned above.

**Comment by tonymet** [comment]
rebased to latest origin/master

**Comment by tonymet** [comment]
Issue #46 -- could be fixed here. 

@nmrshll I'd like to get your thoughts on this one any feedback could help

**Comment by pacoorozco** [comment]
This PR can't be merged due to the changes of the code... and the project. Most of the problems that were rised were either fixed here or fixed on @rfgamaral docker fork.
